Add ABA problem subsection in "Atomic operations as building blocks"

idoleat opened the pull request:
Root causes and an easy example are first provided, and followed by possible solutions. The thread pool example from #7 is reused as more complex example. Not sure if text description of the example is clear enough. If not, figures could then be added.

Currently the thread pool example that fixed the ABA problem looks fine but I am not fully sure if the problem is really fixed. I will try model checking or formal proofing tools to verify that. If it goes well, it may be added as another section on verifying correctness of concurrent programs.

idoleat committed:
Setup the base tone of describing ABA problem

First describing what can go wrong and then explain the root cause of 
it.

Example code will be added later on as well as sultions to ABA problem.

idoleat committed:
Add a simple ABA problem example and more description

Because ABA problem is not so intuitive, the plan is to have a simple
example first, then reusing the new example in section 5. A nother stack
example by "null program" will be mentioned for additional examples.

Next commit will provide solutions to fix ABA problem in both
examples.

idoleat committed:
Add full thread pool example and description

Approaches are explained first, then the code and explanations on
changes.

idoleat committed:
Fix typo and grammar mistakes

idoleat committed:
Fix typo

jserv reviewed (changes requested):
Rebase the latest `main` branch.

idoleat committed:
Update tpool in aba example to the one in main

The new thread pool is ported to the rmw_example_aba.c file. Makefile is
updated accordingly as well.

idoleat committed:
Align tex file with updated code example

The compile command is added in the Makefile, same as the previous 
example code. Missing reference and wrong compiler flag are fixed as 
well.

jserv commented on concurrency-primer.tex:
```diff
@@ -939,7 +939,7 @@ \subsection{ABA problem}
 It could be even worse if nested ABA problem occurs in thread B.
 Also, the possibility to allocate a job with same address could be higher when using memory pool, meaning that more chances to have ABA problem occurred.
-In fact, pre-allocated memory should be used to achive lock-free since \monobox{malloc} could have mutex involved in multithreaded environment.
+In fact, pre-allocated memory should be used to achieve lock-free since \monobox{malloc} could have mutex involved in multithreaded environment.
```
Always use "multi-threaded" instead of "multithreaded."

idoleat replied:
Resolved in 8abe58d.

idoleat committed:
Update terminology from "multithread" to "multi-threaded"

jserv commented on concurrency-primer.tex:
```diff
@@ -883,6 +883,94 @@ \subsection{Conclusion about lock-free}
+
+The execution result would be:
+
+\begin{ccode}
```
The following is not code. Instead, it is plaintext. Don't annotate with `code`.

jserv commented on concurrency-primer.tex:
```diff
@@ -883,6 +883,94 @@ \subsection{Conclusion about lock-free}
 a domain fraught with challenges.
 
+\subsection{ABA problem}
+We have introduced CAS as one of the read-modify-write operations.
```
Proofread as following:
> CAS has been introduced as one of the read-modify-write operations.
> However, the target object not changing does not necessarily mean that no other threads modified it halfway through.
> If the target object is changed by another thread and then changed back, the result of comparison would still be equal. In this case, the target object has indeed been modified, yet the operation appears unchanged, compromising its atomicity.

The above presents them as statements rather than questions.

jserv commented on concurrency-primer.tex:
```diff
@@ -883,6 +883,94 @@ \subsection{Conclusion about lock-free}
+    A: v = 52
+\end{ccode}
+
+In the example provided, the presence of ABA problem results in thread A being unaware that variable \monobox{v} has been altered.
```
Annotate with `\cc` rather than `\monobox`.

jserv commented on concurrency-primer.tex:
```diff
@@ -883,6 +883,94 @@ \subsection{Conclusion about lock-free}
+In fact, pre-allocated memory should be used to achieve lock-free since \monobox{malloc} could have mutex involved in multi-threaded environment.
+
+Failure to recognize changed target object through comparison can result in stale information.
+The general concept of solving this problem involves adding more information to make different state distinguishable, and then making a decision on whether to act on the old state or retry with the new state.
```
This statement was not clear as for the impact of ABA problem. Can you improve it?

jserv commented on concurrency-primer.tex:
```diff
@@ -883,6 +883,94 @@ \subsection{Conclusion about lock-free}
+It requires the compare-and-swap instruction to be capable of comparing a wider size at once.
+Sometimes, this is referred to as \introduce{double-width compare-and-swap}.
+On x86-64 processors, for atomic instructions that load or store more than a CPU word size, it needs additional hardware support.
+You can use \monobox{\$ grep cx16 /proc/cpuinfo} to check if the processor supports 16-byte compare-and-swap.
```
Don't specify `$` for shell commands.

idoleat committed:
Tweak style annotation

Replace \monobox{} with \cc{} to properly apply the style to C code.

Dollar sign removed for shell command.

idoleat committed:
Proofread introduction to the ABA problem

Presents them as statements rather than questions.

idoleat committed:
Improve the statement of the impact of ABA problem

jserv commented:
Thank @idoleat for contributing!
